data/tables/aggregate.js

```diff
@@ -0,0 +1,206 @@
+ * @typedef {import('../types').CarItemAggregate} CarItem
+ */
+
+/** @type {Record<string, AggregateStat>} */
+export const AGGREGATE_STAT = {
+  ingesting: 'INGESTING',
+  ready: 'READY',
+  dealPending: 'DEAL_PENDING',
+  dealProcessed: 'DEAL_PROCESSED'
+}
+
+const DEFAULT_MAX_SIZE = 127*(1<<28)
+const DEFAULT_MIN_SIZE = 1+127*(1<<27)
+
+
+/**
+ * Abstraction layer to handle operations for aggregates to send to Spade.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {AggregateOpts} [options]
+ * @returns {import('../types').AggregateTable}
+ */
+export function createAggregateTable (region, tableName, options = {}) {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint
+  })
+  const minSize = options.minSize || DEFAULT_MIN_SIZE
+  const maxSize = options.maxSize || DEFAULT_MAX_SIZE
+
+  return {
+    /**
+     * Add new CARs to the given aggregate if still with enough space.
+     *
+     * @param {string} aggregateId
+     * @param {CarItem[]} cars 
+     */
+    add: async (aggregateId, cars) => {
+      const links = cars.map(car => car.link)
+      const updateAccumSize = cars.reduce((acc, car) => acc + car.size, 0)
+      const maxSizeBeforeUpdate = maxSize - updateAccumSize
+
+      if (maxSizeBeforeUpdate < 0) {
+        throw new Error('Given cars do not fit inside the given aggregate')
+      }
+
+      const insertedAt = new Date().toISOString()
+
+      const updateItemcommand = new UpdateItemCommand({
+        TableName: tableName,
+        Key: marshall({
+          aggregateId
+        }),
+        ExpressionAttributeValues: {
+          ':insertedAt': { S: insertedAt },
+          ':updatedAt': { S: insertedAt },
+          ':ingestingStat': { S: AGGREGATE_STAT.ingesting },
+          ':initialSize': { N: `${0}` },
+          ':updateAccumSize': { N: `${updateAccumSize}` },
+          ':maxSizeBeforeUpdate': { N: `${maxSizeBeforeUpdate}` },
+          ':links' :{ SS: links } // SS is "String Set"
+        },
+        // Condition expression runs before update. Guarantees that this operation only suceeds:
+        // - when stat is ingesting
+        // - when there is still enough space in the aggregate for this batch size
+        ConditionExpression: `
+        (
+          attribute_not_exists(stat) OR stat = :ingestingStat
+        )
+        AND
+        (
+          attribute_not_exists(size) OR size <= :maxSizeBeforeUpdate
+        )
+        `,
+        // Update row table with:
+        // - insertedAt if row does not exist already
+        // - updatedAt updated with current timestamp
+        // - set state as ingesting if row does not exist already
+        // - increment size
+        UpdateExpression: `
+        SET insertedAt = if_not_exists(insertedAt, :insertedAt),
+          updatedAt = :updatedAt,
+          stat = if_not_exists(stat, :ingestingStat),
+          size = if_not_exists(size, :initialSize) + :updateAccumSize
```

Comment:
```suggestion
        ADD size :updateAccumSize
```

Reply:
`ADD` is for SETs https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html#Expressions.UpdateExpressions.SET

We need to use SET, and also verify if it does not exist (for when this adds the row in first place)

Reply:
I saw:

> **ADD—updating numbers and sets**
> https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html#Expressions.UpdateExpressions.ADD

My reading is that when a number it increments the current value or sets it. Maybe I'm misunderstanding? I just thought it would be slightly easier to read/less code.

Reply:
Ok, you are right. I just tried it 👍🏼  Interesting, so I agree we should do as you suggest